libs/execution_context/ebpf_maps.c

```diff
@@ -915,39 +1148,17 @@ _reap_oldest_map_entry(_In_ ebpf_core_map_t* map)
 
     lru_map = EBPF_FROM_FIELD(ebpf_core_lru_map_t, core_map, map);
 
-    // Walk through all the keys and values and find the oldest one.
-    for (;;) {
-        uint64_t* key_age = NULL;
-        result = ebpf_hash_table_next_key_pointer_and_value(
-            (ebpf_hash_table_t*)lru_map->core_map.data, previous_key, &next_key, (uint8_t**)&value);
-        if (result != EBPF_SUCCESS) {
-            break;
-        }
-
-        key_age = (uint64_t*)_get_value_supplement(map, value);
-        if (*key_age < oldest_key_age) {
-            oldest_key_age = *key_age;
-            oldest_key = next_key;
-        }
-        previous_key = next_key;
-    }
+    // Grab key from the front of the cold list.
+    ebpf_lock_state_t state = ebpf_lock_lock(&lru_map->lock);
+    ebpf_lru_entry_t* entry = EBPF_FROM_FIELD(ebpf_lru_entry_t, list_entry, lru_map->cold_list.Flink);
+    ebpf_lock_unlock(&lru_map->lock, state);
 
-    // If we reached the end of the keys, delete the oldest one found.
-    if (result == EBPF_NO_MORE_KEYS && oldest_key != NULL) {
-        // https://github.com/microsoft/ebpf-for-windows/issues/1741
-        // The delete can fail if the oldest entry has already been deleted. This is a known issue that will be fixed in
-        // a future PR.
-        ebpf_result_t delete_result = ebpf_hash_table_delete((ebpf_hash_table_t*)lru_map->core_map.data, oldest_key);
-        if (delete_result == EBPF_SUCCESS) {
-            return true;
-        }
-    }
-    return false;
+    // Delete may fail if the key was already deleted. Caller should attempt the insert again and if it fails with map
+    // full, then it should call this function again.
+    result = _delete_hash_map_entry(map, entry->key);
+    return result;
 }
 
-static ebpf_result_t
-_delete_hash_map_entry(_In_ ebpf_core_map_t* map, _In_ const uint8_t* key);
-
 static ebpf_result_t
 _find_hash_map_entry(
     _In_ ebpf_core_map_t* map, _In_opt_ const uint8_t* key, bool delete_on_success, _Outptr_ uint8_t** data)
```

Comment:
```suggestion
    _Inout_ ebpf_core_map_t* map, _In_opt_ const uint8_t* key, bool delete_on_success, _Outptr_ uint8_t** data)
```

Reply:
Filed https://github.com/microsoft/ebpf-for-windows/issues/1746 as there are a lot of cases where we should be using _Inout_ or _In_ const.